core/src/main/java/com/novoda/noplayer/internal/exoplayer/NoPlayerAdsLoader.java

```diff
@@ -179,39 +182,57 @@ public void onTimelineChanged(Timeline timeline, @Nullable Object manifest, int
             return;
         }
 
-        // handle only pre-roll start on timeline change
-        if (reason == Player.TIMELINE_CHANGE_REASON_PREPARED) {
-            handleAdvertStart();
-        }
+        updateState();
     }
 
-    private void handleAdvertStart() {
-        if (advertHasNotStarted()) {
-            adGroupIndex = player.getCurrentAdGroupIndex();
-            adIndexInGroup = player.getCurrentAdIndexInAdGroup();
+    private void updateState() {
+        boolean wasPlayingAd = isPlayingAdvert;
+        isPlayingAdvert = player.isPlayingAd();
+
+        int previousAdvertGroupIndex = advertGroupIndex;
+        int previousAdvertIndexInAdvertGroup = advertIndexInAdvertGroup;
 
-            if (canPlayAdverts(adGroupIndex)) {
-                notifyAdvertStart(advertBreaks.get(adGroupIndex));
-            } else {
-                resetAdvertPosition();
-            }
+        advertGroupIndex = isPlayingAdvert ? player.getCurrentAdGroupIndex() : C.INDEX_UNSET;
+        advertIndexInAdvertGroup = isPlayingAdvert ? player.getCurrentAdIndexInAdGroup() : C.INDEX_UNSET;
+        boolean advertFinished = wasPlayingAd && advertIndexInAdvertGroup != previousAdvertIndexInAdvertGroup;
+
+        if (advertFinished) {
+            adPlaybackState = adPlaybackState.withPlayedAd(previousAdvertGroupIndex, previousAdvertIndexInAdvertGroup);
+            updateAdPlaybackState();
+            notifyAdvertEnd(previousAdvertGroupIndex, previousAdvertIndexInAdvertGroup);
+        }
+
+        boolean advertStarted = isPlayingAdvert && advertIndexInAdvertGroup != previousAdvertIndexInAdvertGroup;
+        if (advertStarted) {
+            notifyAdvertStart(advertGroupIndex, advertIndexInAdvertGroup);
         }
     }
 
-    private boolean advertHasNotStarted() {
-        return player.isPlayingAd() && (adGroupIndex == -1 || adIndexInGroup == -1);
+    private void notifyAdvertEnd(int playedAdvertGroupIndex, int playedAdvertIndexInAdvertGroup) {
+        AdvertBreak advertBreak = advertBreaks.get(playedAdvertGroupIndex);
+        List<Advert> adverts = advertBreak.adverts();
```

Comment:
you could add a new method in the `advertBreak.advert(index)` so that you don't have to return the whole list